[FIX] purchase_stock: use currency rate at bill date when invoice before receipt

royle-vietnam opened the pull request:
In some business case when the invoice is done before the receipt, the `stock.valuation.layer` currency rate will be taken on the bill rather than the current date.

To reproduce the issue:
(Need account_accountant)
1. Create a product category PC
   - Costing method: FIFO
   - Inventory valuation: Automated
2. Create a product P
   - Type: Storable
   - Category: PC
3. On 10/11/2024, confirm a PO with 1 x P at Є10
4. On 11/11/2024, bill it with currency rate of 1.06 at $10.6
4. On 12/11/2024, receive P with currency rate of 1.07 at $10.7

Errors: The price SVL is different from the bill

Solution: Use currency rate at bill date rather than the current date when invoice before receipt




---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/187015.png)](https://mergebot.odoo.com/odoo/odoo/pull/187015)

Whenrow commented:
The test test_average_realtime_with_two_delivery_anglo_saxon_valuation_multicurrency_different_dates is failing because you changed the way the convert date is choosen on invoices. It's still not perfectly matching invoices with receipt quantity but I think it's still better. Could you update the test and also add the one written in https://github.com/odoo/odoo/pull/184033 ? Thank you

royle-vietnam commented:
> The test test_average_realtime_with_two_delivery_anglo_saxon_valuation_multicurrency_different_dates is failing because you changed the way the convert date is choosen on invoices. It's still not perfectly matching invoices with receipt quantity but I think it's still better. Could you update the test and also add the one written in #184033 ? Thank you

@Whenrow thanks for your feedback, `test_average_realtime_with_two_delivery_anglo_saxon_valuation_multicurrency_different_dates` is the situation of receiving goods first, invoice later
I have added the condition to make sure it is invoice first, receiving goods later and added the test. Please review again, thanks.

Whenrow commented:
robodoo r+
thank you
